README.md

```diff
@@ -14,52 +14,164 @@ Backend developed by the Digital R&D team for hybrid search apps. Related repos:
+    },
+    {
+        "meta": {"title": "Lighthouse of Alexandria"},
+        "content": "The Lighthouse of Alexandria, sometimes called the Pharos of Alexandria (/ˈfɛərɒs/ FAIR-oss; Ancient Greek: ὁ Φάρος τῆς Ἀλεξανδρείας, romanized: ho Pháros tês Alexandreías, contemporary Koine Greek pronunciation: [ho pʰáros tɛ̂ːs aleksandrěːaːs]; Arabic: فنار الإسكندرية), was a lighthouse built by the Ptolemaic Kingdom of Ancient Egypt, during the reign of Ptolemy II Philadelphus (280–247 BC). It has been estimated to have been at least 100 metres (330 ft) in overall height. One of the Seven Wonders of the Ancient World, for many centuries it was one of the tallest man-made structures in the world.",
+    },
+]
+
+# Put into Haystack Document instances
+docs = [Document(**content) for content in doc_list]
 ```
 
-### Adding new dependencies
+5. Connect to an OpenSearch container set up by `docker compose`
 
-If you need to add a new dependency, insert it into pyproject.toml. Set upper
-and lower bounds for the versions.
+```
+query_document_store = OpenSearchDocumentStore(
+    hosts="http://0.0.0.0:4566/opensearch/eu-west-2/rd-demo",
+    use_ssl=False,
+    verify_certs=False,
+    http_auth=("localstack", "localstack"),
+    embedding_dim=cfg["embedding_dim"],
+    batch_size=cfg["index_batch_size"],
+)
+```
 
-### Start the stack locally
+6. Run the indexing pipeline to write documents to the vector store
 
 ```
-docker compose build
-docker compose up
+indexer = IndexingPipeline(query_document_store, dense_embedding_model=cfg["dense_embedding_model"], semantic=True)
+indexer.index_docs(docs)
+```
+
+7. Set up the retrieval pipeline
+
+You have three options here: (1) BM25 retrieval, (2) dense embedding (semantic) retrieval, (3) hybrid (BM25 + dense embedding) retrieval:
+
+BM25:
+```
+bm25_pipeline = RetrievalPipeline(query_document_store).setup_bm25_pipeline()
+bm25_search_init = Search(bm25_pipeline)
+```
+
+Semantic:
+```
+semantic_pipeline = RetrievalPipeline(
+    query_document_store,
+    dense_embedding_model=cfg['dense_embedding_model'],
+    rerank_model=cfg['rerank_model']
+).setup_semantic_pipeline()
+semantic_search_init = Search(semantic_pipeline)
+```
+
+Hybrid:
+```
+hybrid_pipeline = RetrievalPipeline(
+    query_document_store,
+    dense_embedding_model=cfg['dense_embedding_model'],
+    rerank_model=cfg['rerank_model']
+).setup_hybrid_pipeline()
+hybrid_search_init = Search(hybrid_pipeline)
+```
+
+8. Run a search
+
+BM25:
+```
+test_query = "lighthouse"
+results = bm25_search_init.bm25_search(test_query, top_k=3)
```

Comment:
for clarity, this was in the notebook

Reply:
ah this must just be my expectations - chromadb has a `top_n` param but it'll return all distances (euclidean by default). EG - without setting a distance threshold you can ask for say 10 results and only the closest was any good.

Reply:
Ah so that's a difference between BM25 and semantic search - it should behave as you describe for the semantic/hybrid search options, but BM25 won't return anything that doesn't contain the keywords